Frameworks/CoreGraphics/CGDataProvider.mm

```diff
@@ -16,130 +16,395 @@
+
+public:
+    __RawDataProvider(void* info, const void* data, size_t size, CGDataProviderReleaseDataCallback releaseData)
+        : m_info(info), m_data(data), m_size(size), m_releaseData(releaseData) {
+    }
+
+    const void* GetData() const override {
+        return m_data;
+    }
+
+    size_t GetSize() const override {
+        return m_size;
+    }
+
+    ~__RawDataProvider() {
+        m_releaseData(m_info, m_data, m_size);
+    }
+};
+
+class __DirectDataProvider : public __CGDataProviderInternal {
+    void* m_info;
+    off_t m_size;
+    const CGDataProviderDirectCallbacks* m_callbacks;
+    mutable const void* m_data = nullptr;
+
+public:
+    __DirectDataProvider(void* info, off_t size, const CGDataProviderDirectCallbacks* callbacks)
+        : m_info(info), m_size(size), m_callbacks(callbacks) {
+    }
+    const void* GetData() const override {
+        if (m_data == nullptr) {
+            m_data = m_callbacks->getBytePointer(m_info);
+        }
+
+        return m_data;
+    }
+
+    size_t GetSize() const override {
+        return m_size;
+    }
+
+    ~__DirectDataProvider() {
+        if (m_data != nullptr) {
+            m_callbacks->releaseBytePointer(m_info, m_data);
+        }
+
+        m_callbacks->releaseInfo(m_info);
+    }
+};
+
+class __DirectAccessDataProvider : public __CGDataProviderInternal {
+    void* m_info;
+    size_t m_size;
+    const CGDataProviderDirectAccessCallbacks* m_callbacks;
+    mutable const void* m_data = nullptr;
+
+public:
+    __DirectAccessDataProvider(void* info, size_t size, const CGDataProviderDirectAccessCallbacks* callbacks)
+        : m_info(info), m_size(size), m_callbacks(callbacks) {
+    }
+
+    const void* GetData() const override {
+        if (m_data == nullptr) {
+            m_data = m_callbacks->getBytePointer(m_info);
+        }
+
+        return m_data;
+    }
+
+    size_t GetSize() const override {
+        return m_size;
+    }
+
+    ~__DirectAccessDataProvider() {
+        if (m_data != nullptr) {
+            m_callbacks->releaseBytePointer(m_info, m_data);
+        }
+
+        m_callbacks->releaseProvider(m_info);
+    }
+};
+
+class __SequentialDataProvider : public __CGDataProviderInternal {
+    void* m_info;
+    const CGDataProviderSequentialCallbacks* m_callbacks;
+    mutable std::vector<unsigned char> m_data;
+
+public:
+    __SequentialDataProvider(void* info, const CGDataProviderSequentialCallbacks* callbacks) : m_info(info), m_callbacks(callbacks) {
+    }
+
+    const void* GetData() const override {
+        if (m_data.empty()) {
+            size_t count = 0, delta = 0;
+            do {
+                m_data.resize(m_data.size() + c_streamSize);
+                delta = m_callbacks->getBytes(m_info, m_data.data() + count, c_streamSize);
+                count += delta;
+                m_callbacks->skipForward(m_info, delta);
+            } while (delta == c_streamSize);
+            m_data.resize(count);
```

Comment:
template + function pointer
given most of these classes are pretty much the same with minor difference, it doesn't make since to duplicate code.